core/src/main/java/org/infinispan/xsite/statetransfer/SyncProviderState.java

```diff
@@ -0,0 +1,239 @@
+
+   private final XSiteBackup backup;
+   private final XSiteStateTransferConfiguration configuration;
+   private volatile OutboundTask task;
+
+   private SyncProviderState(XSiteBackup backup, XSiteStateTransferConfiguration configuration) {
+      this.backup = backup;
+      this.configuration = configuration;
+   }
+
+   public static SyncProviderState create(BackupConfiguration config) {
+      XSiteBackup backup = new XSiteBackup(config.site(), true, config.stateTransfer().timeout());
+      return new SyncProviderState(backup, config.stateTransfer());
+   }
+
+   @Override
+   public XSiteStatePushTask createPushTask(Address originator, XSiteStateProvider provider) {
+      OutboundTask newTask = new OutboundTask(originator, provider, this);
+      return TASK_UPDATER.compareAndSet(this, null, newTask) ? newTask : null;
+   }
+
+   @Override
+   public void cancelTransfer() {
+      OutboundTask currentTask = TASK_UPDATER.getAndSet(this, null);
+      if (currentTask != null) {
+         currentTask.cancel();
+      }
+   }
+
+   @Override
+   public boolean isSending() {
+      return task != null;
+   }
+
+   @Override
+   public boolean isOriginatorMissing(Collection<Address> members) {
+      OutboundTask currentTask = task;
+      return currentTask != null && !members.contains(currentTask.coordinator);
+   }
+
+   // methods for OutboundTask
+   void taskFinished() {
+      TASK_UPDATER.set(this, null);
+   }
+
+   XSiteBackup getBackup() {
+      return backup;
+   }
+
+   int getChunkSize() {
+      return configuration.chunkSize();
+   }
+
+   long getWaitTimeMillis() {
+      return configuration.waitTime();
+   }
+
+   int getMaxRetries() {
+      return configuration.maxRetries();
+   }
+
+   private static class OutboundTask implements XSiteStatePushTask, Predicate<List<XSiteState>>, Function<List<XSiteState>, CompletableSource>, CompletableObserver {
+
+      private final Address coordinator;
+      private final XSiteStateProvider provider;
+      private final SyncProviderState state;
+      private volatile boolean canceled = false;
+
+      private OutboundTask(Address coordinator, XSiteStateProvider provider, SyncProviderState state) {
+         this.coordinator = coordinator;
+         this.provider = provider;
+         this.state = state;
+      }
+
+      @Override
+      public void execute(Flowable<XSiteState> flowable, CompletionStage<Void> delayer) {
+         //delayer is the cache topology future. we need to ensure the topology id is installed before iterating
+         delayer.thenRunAsync(() -> flowable
+                     .buffer(state.getChunkSize())
+                     .takeUntil(this)
+                     .concatMapCompletable(this, 1)
+                     .subscribe(this),
+               provider.getExecutor());
```

Comment:
Is this executor really needed? I would think the concatMap operation should be non blocking.

Reply:
really needed? no.
But is it better to do the iteration in another thread. The `delayer` future is completed in `org.infinispan.statetransfer.StateTransferLockImpl#notifyTopologyInstalled` when the node is installing the `CacheTopology`. I don't want to start iterating the data container entries while there is pending work to finish the `CacheTopology` update.

Reply:
Ah okay I wasn't aware this is invoked during a topology update. Then I agree to make sure it completes as quick as possible offloading to another thread is good.